File: src/application/useCases/loadTransactionsScreen.ts
import { merchantIndex } from '../../domain/merchantIndex'
import { normalizeText } from '../../domain/normalize'
import { computePeriodTotals } from '../../domain/ledger'
import { withEstimatedKinds } from '../../domain/estimatedKinds'
import { savingsRatePercent } from '../../domain/formatMoney'
import { buildPeriod, formatPeriodRange, periodForDate, type Period } from '../../domain/period'
import type { Category, Transaction } from '../../domain/entities/types'
import type {
  AllocationRepository,
  CategoryRepository,
  MerchantRepository,
  TagRepository,
  TransactionTagRepository,
  TransactionRepository,
} from '../ports/repositories'

/**
 * LoadTransactionsScreen — تجهيز كل ما تحتاجه شاشة العمليات.
 *
 * وجودها هنا هو ما يحقق قاعدة ARCHITECTURE.md §3:
 * الشاشة لا تكلّم مستودعًا ولا تحسب مبلغًا. كل الحساب هنا وفي domain/.
 *
 * القراءة **محدودة بفترة** — ARCHITECTURE.md §5.6.
 */

export interface TransactionsScreenData {
  period: Period
  periodRange: string
  transactions: Transaction[]
  categories: Category[]
  tagNamesByTransaction: Record<string,string[]>
  merchantNamesByTransaction: Record<string,string[]>

  /**
   * المبالغ **قابلة لأن تكون null**، وهذا مقصود.
   *
   * CLAUDE.md #10: «لا رقم بلا مصدر. لا صفر مؤكد مكان المجهول.»
   *
   * العملية المستوردة تبدأ بنوع اقتصادي غير محدد لأن spec/02 يمنع استنتاجه
   * من اتجاه السيولة. فلو كانت كل عمليات الفترة غير محددة النوع، فالدخل
   * والمصروف **مجهولان لا صفران** — عرض 0.00 وقتها كذب على المستخدم.
   */
  incomeMinor: number | null
  expenseMinor: number | null
  remainingMinor: number | null
  savingsRatePercent: number | null

  /** عدد العمليات التي لم يُحدَّد نوعها الاقتصادي بعد. */
  unclassifiedCount: number
  /** كام عملية اتحسبت بنوع تقديري (OVERRIDES §18). */
  estimatedCount: number
  /** منها: كام عملية التطبيق مش متأكد منها — محتاجة تأكيد. */
  needsReviewCount: number
  /** عدد عمليات الفترة كلها. */
  totalCount: number
}

export interface LoadTransactionsScreenDeps {
  txns: TransactionRepository
  categories: CategoryRepository
  allocations: AllocationRepository
  merchants?: MerchantRepository
  tags?: TagRepository
  transactionTags?: TransactionTagRepository
}

export function makeLoadTransactionsScreen(deps: LoadTransactionsScreenDeps) {
  return async function load(options: {
    period?: Period
    today?: string
    payday?: number
  }): Promise<TransactionsScreenData> {
    const period =
      options.period ??
      (options.today
        ? periodForDate(options.today, options.payday)
        : buildPeriod(new Date().getUTCFullYear(), new Date().getUTCMonth() + 1, options.payday))

    const [transactions, categories] = await Promise.all([
      deps.txns.listByDateRange(period.start, period.end),
      deps.categories.listAll(),
    ])

    const ids = transactions.map(t=>t.id)
    const [allocations,tags,links] = await Promise.all([
      deps.allocations.listByTransactionIds(ids),deps.tags?.listAll()??[],deps.transactionTags?.listByTransactionIds(ids)??[]])
    const tagNames = new Map(tags.map(t=>[t.id,t.displayName]))
    const tagNamesByTransaction: Record<string,string[]> = {}
    for (const link of links) {
      const name = tagNames.get(link.tagId)
      if (name) tagNamesByTransaction[link.transactionId] = [...new Set([...(tagNamesByTransaction[link.transactionId]??[]),name])]
    }
    // المجاميع بالأنواع التقديرية للواضح (OVERRIDES §18)؛ القائمة المعروضة تفضل بالعمليات الأصلية
    const estimated = withEstimatedKinds(transactions, new Map(categories.map((c) => [c.id, c.name])))
    const totals = computePeriodTotals(estimated.transactions, allocations)

    const merchantMap=merchantIndex(await deps.merchants?.listAll()??[])
    const merchantNamesByTransaction:Record<string,string[]>={}
    for(const t of transactions) {
      const m=merchantMap.get(normalizeText(t.rawMerchantName??''))
      if(m) merchantNamesByTransaction[t.id]=[m.displayName,...(m.aliases??[])]
    }
    // «محتاجة تحديد نوع» = اللي التطبيق مش متأكد منها؛ المجاميع نفسها متاحة دايمًا (OVERRIDES §18)
    const unclassifiedCount = estimated.needsReviewCount
    const totalCount = transactions.length

    // فيه عمليات، لكن ولا واحدة محددة النوع ⇒ المجاميع مجهولة لا صفر
    const allUnknown = totalCount > 0 && unclassifiedCount === totalCount

    return {
      period,
      periodRange: formatPeriodRange(period),
      tagNamesByTransaction,
      merchantNamesByTransaction,
      transactions,
      categories,
      incomeMinor: allUnknown ? null : totals.incomeMinor,
      expenseMinor: allUnknown ? null : totals.personalExpenseMinor,
      remainingMinor: allUnknown ? null : totals.remainingMinor,
      savingsRatePercent: allUnknown
        ? null
        : savingsRatePercent(totals.incomeMinor, totals.remainingMinor),
      unclassifiedCount,
      estimatedCount: estimated.estimatedCount,
      needsReviewCount: estimated.needsReviewCount,
      totalCount,
    }
  }
}

File: native-app/app/src/commonMain/kotlin/app/masroufy/usecase/LoadTransactionsScreen.kt
package app.masroufy.usecase

import app.masroufy.core.Category
import app.masroufy.core.Halalas
import app.masroufy.core.Id
import app.masroufy.core.Period
import app.masroufy.core.Transaction
import app.masroufy.core.computePeriodTotals
import app.masroufy.core.formatPeriodRange
import app.masroufy.core.merchantIndex
import app.masroufy.core.normalizeText
import app.masroufy.core.periodForDate
import app.masroufy.core.savingsRatePercent
import app.masroufy.core.withEstimatedKinds
import app.masroufy.port.AllocationRepository
import app.masroufy.port.CategoryRepository
import app.masroufy.port.MerchantRepository
import app.masroufy.port.TagRepository
import app.masroufy.port.TransactionRepository
import app.masroufy.port.TransactionTagRepository

/**
 * شاشة العمليات — نقل `src/application/useCases/loadTransactionsScreen.ts`.
 * الشاشة ما بتكلمش مستودع ولا بتحسب مبلغ (ARCHITECTURE §3)، والقراية محدودة بفترة (§5.6).
 *
 * المبالغ ممكن تكون `null` عن قصد — CLAUDE.md #10: «لا صفر مؤكد مكان المجهول».
 */

data class TransactionsScreenData(
    val period: Period,
    val periodRange: String,
    val transactions: List<Transaction>,
    val categories: List<Category>,
    /** أسماء الوسوم لكل عملية — الاسم المعروض، من غير تكرار وبترتيب أول ظهور. */
    val tagNamesByTransaction: Map<Id, List<String>>,
    /** اسم التاجر الأساسي وبعده البديلة — للبحث والعرض. */
    val merchantNamesByTransaction: Map<Id, List<String>>,
    val incomeMinor: Halalas?,
    val expenseMinor: Halalas?,
    val remainingMinor: Halalas?,
    val savingsRatePercent: Double?,
    /** «محتاجة تحديد نوع» = اللي التطبيق مش متأكد منها؛ المجاميع نفسها متاحة دايمًا (OVERRIDES §18). */
    val unclassifiedCount: Int,
    /** كام عملية اتحسبت بنوع تقديري (OVERRIDES §18). */
    val estimatedCount: Int,
    /** منهم: كام عملية التطبيق مش متأكد منها — محتاجة تأكيد. */
    val needsReviewCount: Int,
    val totalCount: Int,
)

data class LoadTransactionsScreenDeps(
    val txns: TransactionRepository,
    val categories: CategoryRepository,
    val allocations: AllocationRepository,
    val merchants: MerchantRepository? = null,
    val tags: TagRepository? = null,
    val transactionTags: TransactionTagRepository? = null,
)

/**
 * التطبيق الحالي لو مفيش فترة ولا يوم بياخد يوم الجهاز. هنا الوقت بييجي من برّه دايمًا
 * (نفس قاعدة `Clock` port)، فلازم `period` أو `today` — والشاشة بتبعت واحد منهم دايمًا.
 */
data class LoadTransactionsScreenRequest(
    val period: Period? = null,
    val today: String? = null,
    val payday: Int? = null,
)

class LoadTransactionsScreen(private val deps: LoadTransactionsScreenDeps) {
    suspend fun load(request: LoadTransactionsScreenRequest): TransactionsScreenData {
        val period = request.period
            ?: request.today?.let { if (request.payday != null) periodForDate(it, request.payday) else periodForDate(it) }
            ?: throw IllegalArgumentException("لازم فترة أو تاريخ اليوم — مفيش ساعة جوه المنطق")

        val transactions = deps.txns.listByDateRange(period.start, period.end)
        val categories = deps.categories.listAll()

        val ids = transactions.map { it.id }
        val allocations = deps.allocations.listByTransactionIds(ids)
        val tags = deps.tags?.listAll() ?: emptyList()
        val links = deps.transactionTags?.listByTransactionIds(ids) ?: emptyList()

        // الرابط اللي وسمه مش موجود بيتساب، والاسم المكرر لنفس العملية بيتشال بترتيب أول ظهور
        val tagNames = tags.associate { it.id to it.displayName }
        val tagNamesByTransaction = LinkedHashMap<Id, List<String>>()
        for (link in links) {
            val name = tagNames[link.tagId] ?: continue
            val merged = LinkedHashSet(tagNamesByTransaction[link.transactionId].orEmpty())
            merged.add(name)
            tagNamesByTransaction[link.transactionId] = merged.toList()
        }

        // المجاميع بالأنواع التقديرية للواضح (OVERRIDES §18)؛ القايمة المعروضة بتفضل بالعمليات الأصلية
        val estimated = withEstimatedKinds(transactions, categories.associate { it.id to it.name })
        val totals = computePeriodTotals(estimated.transactions, allocations)

        val merchantMap = merchantIndex(deps.merchants?.listAll() ?: emptyList())
        val merchantNamesByTransaction = LinkedHashMap<Id, List<String>>()
        for (t in transactions) {
            val merchant = merchantMap[normalizeText(t.rawMerchantName ?: "")] ?: continue
            merchantNamesByTransaction[t.id] = listOf(merchant.displayName) + merchant.aliases.orEmpty()
        }

        val unclassifiedCount = estimated.needsReviewCount
        val totalCount = transactions.size

        // فيه عمليات، لكن ولا واحدة اتحدد نوعها ⇒ المجاميع مجهولة مش صفر
        val allUnknown = totalCount > 0 && unclassifiedCount == totalCount

        return TransactionsScreenData(
            period = period,
            periodRange = formatPeriodRange(period),
            transactions = transactions,
            categories = categories,
            tagNamesByTransaction = tagNamesByTransaction,
            merchantNamesByTransaction = merchantNamesByTransaction,
            incomeMinor = if (allUnknown) null else totals.incomeMinor,
            expenseMinor = if (allUnknown) null else totals.personalExpenseMinor,
            remainingMinor = if (allUnknown) null else totals.remainingMinor,
            savingsRatePercent = if (allUnknown) null else savingsRatePercent(totals.incomeMinor, totals.remainingMinor),
            unclassifiedCount = unclassifiedCount,
            estimatedCount = estimated.estimatedCount,
            needsReviewCount = estimated.needsReviewCount,
            totalCount = totalCount,
        )
    }
}
